### PR TITLE
Overhaul our handling of GitHub labels.

### PR DESCRIPTION
**Pull request by Ferroin**
### Summary

This is a comprehensive update of how we are handling GitHub issue and PR labels. I was originally going to write up a proposal suggesting these changes, but I realized part way through doing so that that was going to take about as much effort as it would to just implement the changes and open a draft PR for discussion, and this approach should result in much less bikeshedding than trying to sort out a concrete plan before implementing things would.

#### Labels as Code

The key part of this PR involves establishing all of our labels as actual code in the repository itself. The label configuration is stored in a file called `.github/labels.yml`, and this configuration is applied to the repo by a new GHA workflow located at `.github/workflows/label-sync.yml`.  Handling GitHub labels in this way has a number of specific benefits for the long-term manageability of labels in the repository:

- Changes to labels will involve a commit to the repo, which ensures that they are at least partially documented in the form of a commit indicating when they were changed.
- We can properly describe _groups_ of labels in this configuration file, instead of having to describe each label individually.
- We can enforce sanity checks on the label configuration (see the section below on Label Config Validation for more about this).

#### Removed Labels

The label config added by this PR _removes_ a number of labels that are currently present in our repository:

- Labels that should be project views instead: `Customer Success`, `config. manager`, `distributed auth`, `Use Cases`, and all of the `*-team` labels (which should not only be project views, but should also involve assigning appropriate team members on the issues.
- Labels that are no longer being used at all: `area/backends`, `javascript`, `monitoring discussion`, `Q1 GOAL`, `ready`, `sso changes`, `synthetic charts`.
- Labels that are not actually particularly useful (almost no issues using them, said issues generally share a lot of other labels, and the differentiation is not hugely relevant for us): `community-feedback`, `discussion`, `enabler`, `Flag`, `internal`, `Roadmap`.
- `area/ux`: This one is a special case, any issue that you _think_ needs this label should be tagged with the correct area label for what part of the software is affected instead. If a comprehensive view of this is actually required, then a project view should be used instead.
- `blocked`: Again, special case, anything that’s blocked should be labeled based on _why_ it’s blocked instead. If a comprehensive view of this is actually required, then a project should be used instead.
- `area/web`: Has been split into explicit `area/web/api`, `area/web/gui`, and `area/web/server` labels, as the three parts are functionally very distinct components.

#### Renamed Labels

The following labels have been renamed to ensure consistency:

- `area/ACLK` (this is live already: Renamed to `area/aclk` for consistency with our other `area/*` labels.
- `area/collectors`: Functionally replaced by existing `collectors/*` labels for specific collectors.
- `area/docs`: Renamed to `documenttion` for consistency with most other GitHub projects.
- `area/packaging/third-party`: Merged into the `upstream bug` label.
- `area/upstream`: Renamed to `upstream bug` to clearly identify the purpose of the label.
- `area/webapi`: Renamed to `area/web/api`.
- `area/windows`: Renamed to `platform/windows`.
- `docker`: Renamed to `platform/docker`.
- `K8s`: Renamed to `platform/k8s`.
- `os/bsd`: Renamed to `platform/freebsd`
- `os/macos`: Renamed to `platform/macos`
- `new collector`: Renamed to `collectors/new`.
- `run-ci/packaging`: Renamed to `run-ci/full-packaging` to better reflect what it does.

I intend to manually handle all the renames prior to merging this PR to preserve continuity of the labels.

#### Added Labels

This PR also adds a number of new labels:

- A new `priority/critical` label has been added for things that need to be fixed _right now_ (most other projects that do priority labels on GitHub use four priority categories, and this also happens to be consistent with our issue templates).
- New `area/packaging/deb` and `area/packaging/rpm` labels have been added specifically for things that relate to the DEB or RPM packaging code, to better differentiate from general platform support issues.
- A series of `platform/linux` labels have been added to tag issues relating to platform support for specific Linux distributions.
- A new `platform/new` label has been added to tag issues relating to adding support for new platforms (or Linux distributions). 

#### Label Config Validation 

One of the big advantages of labels as code is that we can validate our label configuration based on other aspects of our code. To that end, this PR adds a new CI job that does just that. The following aspects of the label configuration are validated by this job:

- The syntax of `.github/labels.yml` is correct.
- All labels have a description (this is important for third-party contributors).
- The syntax of `.github/labeler.yml` (the PR auto-labeler configuration) is correct.
- All labels used by `.github/labeler.yml` are also defined in `.github/labels.yml`.
- All `area/*` labels have an associated source directory in the source tree unless they specifically should not.
- All `collectors/*` labels have an associated source directory under `collectors/` unless they specifically should not.
- All collectors have an associated `collectors/*` label.
- All `platform/*` labels are properly in sync with `.github/data/distros.yml` (the master file defining our CI build matrices).

#### Other Changes

In addition to the above changes, this also sorts the labels in our PR auto-labeler configuration correctly, and adds a new rule to automatically tag PRs that add a new directory under `collectors/` with the `collectors/new` label.

### Test Plan

n/a

### Additional Information

@netdata/agent, @netdata/product This obviously needs discussion before being merged. I’ve tagged you all for review (as well as explicitly tagging @cakrit and @ktsaou so they see this too) in the hopes of having actual discussion on this. I plan to keep this PR listed as a Draft to prevent accidental merges until we’ve decided for certain how we want all of this to work, but would still like review feedback on it as it is now (the only likely changes are what labels are defined and what labels are not defined).

### CODE REVIEW

**Comment by ilyam8** [comment]
> area/collectors: Functionally replaced by existing collectors/* labels for specific collectors.

It is not. Few examples of how I use it:
 - an issue with a collector: area/collectors + collectors/X.
 - an issue with an alarm for the collector: area/health + collectors/X.
 - an issue with a collector docs: area/docs + collectors/X.

**Comment by ilyam8** [comment]
It is very desirable to not delete it, the suggested approach (_the presence of collectors/* and no other associated area tag_) doesn't work for me:

- as I said collector/X is not an area and I use it to narrow down an issue (e.g. packaging + collector/X, health + collector/x, docs + collector/X, collectors + collector/x).
- I use labels for filtering issues and area/collectors allow me to see everything related to collectors. I think effective filtering is top1  priority. If our set of labels doesn't provide it we need to adjust it.

**Comment by Ferroin** [comment]
> It is very desirable to not delete it, the suggested approach (_the presence of collectors/* and no other associated area tag_) doesn't work for me:
> 
> * as I said collector/X is not an area and I use it to narrow down an issue (e.g. packaging + collector/X, health + collector/x, docs + collector/X, collectors + collector/x).
> * I use labels for filtering issues and area/collectors allow me to see everything related to collectors. I think effective filtering is top1  priority. If our set of labels doesn't provide it we need to adjust it.

Point taken, I’ve re-added it to the label config.

**Comment by andrewm4894** [comment]
@Ferroin how will this effect existing issues? Would we need some sort of script to relabel them based on new mapping etc? or is that somehow something that can be done automatically as part of this PR and GH actions etc?

**Comment by Ferroin** [comment]
> @Ferroin how will this effect existing issues? Would we need some sort of script to relabel them based on new mapping etc? or is that somehow something that can be done automatically as part of this PR and GH actions etc?

I plan to manually process the renames (those are easy, I can just rename the existing labels and any issues with those labels will be automatically updated) prior to merging the PR.

The removals are a bit trickier. `area/web` will need manual handling. The others need double-checked, but in theory should not need any special handling.

**Comment by thiagoftsm** [comment]
@Ferroin are you going to address [this](https://app.codacy.com/gh/netdata/netdata/pullRequest?prid=11305941) codacy warning.

File: .github/labels.yml [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+  description: Issues and PRs related to the xenstat data collection plugin
+  color: *collectors
+
+# CI Override Labels
+#
+# These are interpreted by our CI when seen on PRs
+# Usually, they force running some CI that would not normally run on that PR
+- name: run-ci/actionlint
+  description: Force running actionlint against labeled PRs
+  color: &run-ci b1a6fb
```
Comment (thiagoftsm) [inline .github/labels.yml:291]: This is the unique color defined with `&run-ci`, is this a typo error?

Comment (Ferroin) [reply .github/labels.yml:291]: Nope, it’s defining a YAML anchor with that exact value, which is then referenced in every other `run-ci` label’s color definition.

**Comment by Ferroin** [comment]
> @Ferroin are you going to address [this](https://app.codacy.com/gh/netdata/netdata/pullRequest?prid=11305941) codacy warning.

Codacy still hasn’t learned about PEP 634 structural pattern matching yet, so it’s still throwing false positives like this.

**Review by thiagoftsm — APPROVED** [review]
Considering the answers I and other teammates received,  LGTM!

**Comment by hugovalente-pm** [comment]
> Labels that should be project views instead: Customer Success, config. manager, distributed auth, Use Cases, and all of the *-team labels (which should not only be project views, but should also involve assigning appropriate team members on the issues.

I'm not sure I fully get the suggestion, for example I know the `Use Cases` label is driven to consolidate all the work for this project across multiple repos https://github.com/orgs/netdata/projects/47/views/1
Not sure there is an automation to automatically add tickets to this board if the ticket is labelled with it but probably @cakrit @sashwathn or @shyamvalsan can chime in on that but trying to say a project view/board is being created based on the assumption of that label being there.

Also with the changes happening on how we are structuring things not sure that board will continue to be used, probably yes as the main backlog of thins identified for the Use Cases team where just a few will probably make it to the [Product initiatives wide view](https://github.com/orgs/netdata/projects/15/views/12)

>  *-team labels (which should not only be project views, but should also involve assigning appropriate team members on the issues.

for this specifically, this was when we had the Vertical teams (Visualizations, Alerts, Mgmnt and Navigation, etc.) but since these are not longer there maybe they can be removed but having the label didn't actually mean to have a person assigned to the ticket since the work could still be in the backlog and not being actively worked on due to other priorities

in summary, from Product pov I can't assume this is ok to be merged without getting other people's inputs

**Comment by Ferroin** [comment]
> > Labels that should be project views instead: Customer Success, config. manager, distributed auth, Use Cases, and all of the *-team labels (which should not only be project views, but should also involve assigning appropriate team members on the issues.
> 
> I'm not sure I fully get the suggestion, for example I know the `Use Cases` label is driven to consolidate all the work for this project across multiple repos https://github.com/orgs/netdata/projects/47/views/1 Not sure there is an automation to automatically add tickets to this board if the ticket is labelled with it but probably @cakrit @sashwathn or @shyamvalsan can chime in on that but trying to say a project view/board is being created based on the assumption of that label being there.

I may also not be fully understanding how GitHub projects work here, but from the perspective of external users these labels look like they were just thrown around randomly on a few issues and then forgotten about. Consider the fact that I’m an employee who has worked here since before these labels were first added, and I still had no idea what they were about.

If we _really_ need to keep them, then they should use a shared prefix identifying that they are project labels so that we can avoid this kind of confusion in the future.

> Also with the changes happening on how we are structuring things not sure that board will continue to be used, probably yes as the main backlog of thins identified for the Use Cases team where just a few will probably make it to the [Product initiatives wide view](https://github.com/orgs/netdata/projects/15/views/12)
> 
> > *-team labels (which should not only be project views, but should also involve assigning appropriate team members on the issues.
> 
> for this specifically, this was when we had the Vertical teams (Visualizations, Alerts, Mgmnt and Navigation, etc.) but since these are not longer there maybe they can be removed but having the label didn't actually mean to have a person assigned to the ticket since the work could still be in the backlog and not being actively worked on due to other priorities

For cases like that, I would argue that assigning all of the team members is the right thing to do. In an ideal world GitHub would let us assign teams directly to issues (people have been asking for this for years now, GitHub seems to not care for unknown reasons despite this being how most large projects actually work), but based on personal experience unless the team is more than about half a dozen people, this approach works reasonably well.

The part that makes these cases tricky is that Netdata has so many repos and we’re trying to keep things in sync across those repos (though we could do that _much_ more easily using labels as code, just have a workflow in each repo pulling from one master location that defines all the labels), which leads to confusion when we have labels like this that look borderline useless even to internal contributors.

> in summary, from Product pov I can't assume this is ok to be merged without getting other people's inputs

Which is exactly why I’m seeking feedback.

**Comment by cakrit** [comment]
In general, I like all the suggestions here, though I'm not quite sure about moving label management to a file. The Github UI to add and manage labels will still be available @Ferroin? How will that work? 

I agree that project-related labels are no longer needed, including the "use cases" label. Github projects should be sufficient to organize the information.

Having said that, I personally wouldn't want us to become "label Nazis" and strictly gatekeep this. If a label is useful for a certain amount of time, to some people, I don't see why others would need to approve whether or not it should be added.

**Comment by Ferroin** [comment]
> In general, I like all the suggestions here, though I'm not quite sure about moving label management to a file. The Github UI to add and manage labels will still be available @Ferroin? How will that work?

Technically yes, but as things stand right now any changes there would be overwritten whenever the file sync runs. We can technically disable that, but doing so also means that we lose 99% of the benefits of having labels managed as code instead of repository metainformation (because pretty much all of them rely on the file being in-sync with the actual labels on the repo).

I understand that opening a PR to add a label slows things down a bit, but it also lets enforce common sense things like making sure that the label actually has a proper description explaining what it’s for, as well as adding other sanity checks to make sure we don’t end up with dead labels like `area/backends` again. It also provides a clear location to handle inconsistencies in labeling. I seriously doubt anybody would prefer getting random emails from CI complaining about a label not having a description, and I can guarantee that such emails would not result in _any_ enforcement of such a rule.

> I agree that project-related labels are no longer needed, including the "use cases" label. Github projects should be sufficient to organize the information.
> 
> Having said that, I personally wouldn't want us to become "label Nazis" and strictly gatekeep this. If a label is useful for a certain amount of time, to some people, I don't see why others would need to approve whether or not it should be added.

The goal is not so much that we should be gatekeeping labels, but that we should make sure that they are actually understandable and useful, not cryptic things where even our own employees aren’t certain about what they are for, and that we have a proper indication of who added a given label and why (which in turn helps with cleaning up old labels).

As far as approvals, if the label is actually useful to some group of people, then that group of people should approve the PR to add that label. The required approvals will indirectly ensure that labels that get added _are_ useful to some group of people, and not just something that one person thinks we need.

**Comment by cakrit** [comment]
I'm personally happy with the responses, no objections from me.
